src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,397 @@
+                        }
+                        // REVIEW: Can this be faster?
+                        if (ContainsChar(memory.Span[ichCur], separators))
+                            break;
+                    }
+
+                    yield return memory.Slice(ichMinLocal, ichCur - ichMinLocal);
+
+                    // Skip the separator.
+                    ichCur++;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Splits <paramref name="memory"/> on the left-most occurrence of separator and produces the left
+        /// and right <see cref="ReadOnlyMemory{T}"/> of <see cref="char"/> values. If <paramref name="memory"/> does not contain the separator character,
+        /// this returns false and sets <paramref name="left"/> to this instance and <paramref name="right"/>
+        /// to the default <see cref="ReadOnlyMemory{T}"/> of <see cref="char"/> value.
+        /// </summary>
+        public static bool SplitOne(char separator, out ReadOnlyMemory<char> left, out ReadOnlyMemory<char> right, ReadOnlyMemory<char> memory)
+        {
+            if (memory.IsEmpty)
+            {
+                left = memory;
+                right = default;
+                return false;
+            }
+
+            var span = memory.Span;
+            int ichCur = 0;
+            for (; ; ichCur++)
+            {
+                Contracts.Assert(0 <= ichCur && ichCur <= memory.Length);
+                if (ichCur >= memory.Length)
+                {
+                    left = memory;
+                    right = default;
+                    return false;
+                }
+                if (span[ichCur] == separator)
+                    break;
+            }
+
+            left = memory.Slice(0, ichCur);
+            right = memory.Slice(ichCur + 1, memory.Length - ichCur - 1);
+            return true;
+        }
+
+        /// <summary>
+        /// Splits <paramref name="memory"/> on the left-most occurrence of an element of separators character array and
+        /// produces the left and right <see cref="ReadOnlyMemory{T}"/> of <see cref="char"/> values. If <paramref name="memory"/> does not contain any of the
+        /// characters in separators, this return false and initializes <paramref name="left"/> to this instance
+        /// and <paramref name="right"/> to the default <see cref="ReadOnlyMemory{T}"/> of <see cref="char"/> value.
+        /// </summary>
+        public static bool SplitOne(char[] separators, out ReadOnlyMemory<char> left, out ReadOnlyMemory<char> right, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValueOrNull(separators);
+
+            if (memory.IsEmpty || separators == null || separators.Length == 0)
+            {
+                left = memory;
+                right = default;
+                return false;
+            }
+
+            var span = memory.Span;
+
+            int ichCur = 0;
+            if (separators.Length == 1)
```

Comment:
Use `MemoryExtensions.IndexOfAny` instead?

Reply:
These take Span as input and in our case we have character(s) as input so I will have to convert character to span which I think is expensive because AsMemory() seemed expensive so may be AsSpan might also be?

---
In reply to: [216468683](https://github.com/dotnet/machinelearning/pull/863#discussion_r216468683) [](ancestors = 216468683)

Reply:
You have a `char[] separators` which is simply converted to Span by calling `.AsSpan()` on it. You won't be doing it in a loop, so it won't incur a perf penalty.  Plus IndexOfAny is getting vectorized in .NET Core 3.0.  See [https://github.com/dotnet/coreclr/pull/19790](https://github.com/dotnet/coreclr/pull/19790)

---
In reply to: [216845408](https://github.com/dotnet/machinelearning/pull/863#discussion_r216845408) [](ancestors = 216845408,216468683)